p2p-tests/lib/p2p.ts

```diff
@@ -104,66 +86,151 @@ export class LinePrefix extends stream.Transform {
   }
 }
 
-export function initPeer(namespace: string): void {
-  const radHome = workspacePath([namespace]);
-  fs.removeSync(radHome);
-  fs.mkdirsSync(radHome);
-
-  execa.sync(path.join(BIN_PATH, "radicle-proxy-init"), [
-    namespace,
-    "--key-passphrase",
-    namespace,
-    "--rad-home",
-    workspacePath([namespace]),
-  ]);
+interface PeerParams {
+  namespace: string;
+  ipAddress: string;
+  startArgs: string[];
+  dataPath: string;
 }
 
-export function startPeer(
-  namespace: string,
-  ip: string,
-  args: string[]
-): ProxyClient.ProxyClient {
-  initPeer(namespace);
-
-  const childProcess = spawnInNamespace(
-    namespace,
-    [
-      path.join(BIN_PATH, "radicle-proxy"),
-      "--peer-listen",
-      `${ip}:8776`,
-      "--http-listen",
-      `${ip}:17246`,
-      "--skip-remote-helper-install",
-      "--unsafe-fast-keystore",
-      "--insecure-http-api",
-      "--dev-log",
-      ...args,
-    ],
-    {
-      RAD_HOME: workspacePath([namespace]),
+class Peer {
+  public dataPath: string;
+  public radHome: string;
+  public namespace: string;
+  public ipAddress: string;
+  public startArgs: string[];
+
+  _childProcess: execa.ExecaChildProcess | undefined = undefined;
+
+  constructor({ namespace, ipAddress, startArgs, dataPath }: PeerParams) {
+    this.dataPath = dataPath;
+    this.radHome = path.join(dataPath, `${namespace}-rad-home`);
+    this.namespace = namespace;
+    this.ipAddress = ipAddress;
+    this.startArgs = startArgs;
+
+    fs.mkdirsSync(this.radHome);
+  }
+
+  get childProcess() {
+    if (this._childProcess) {
+      return this._childProcess;
+    } else {
+      throw new Error(
+        "Tried to access childProcess before it was initialized."
+      );
     }
-  );
+  }
 
-  processes.push(childProcess);
+  async stop(): Promise<void> {
+    if (!this.childProcess) {
+      throw new Error("Tried to stop() process that wasn't started.");
+    }
+
+    this.childProcess.kill("SIGTERM");
+    await this.childProcess;
+  }
 
-  return new ProxyClient.ProxyClient(`http://${ip}:17246`);
+  async kill(): Promise<void> {
+    if (!this.childProcess) {
+      throw new Error("Tried to kill() process that wasn't started.");
+    }
+
+    try {
+      this.childProcess.kill("SIGKILL");
+      await this.childProcess;
+    } catch {
+      // Ignore errors.
+    }
+  }
 }
 
-export function startSeed(namespace: string, args: string[]): void {
-  const radHome = workspacePath([namespace]);
-  fs.removeSync(radHome);
-  fs.mkdirsSync(radHome);
+export class UpstreamPeer extends Peer {
```

Comment:
Should be called `UpstreamProxy` or `RadicleProxy` to align it with the name of the binary.

Reply:
96baf17dd531225772299943f80ba7908bdaa278